lib/random_tools.rb

```diff
@@ -0,0 +1,40 @@
+module Upaya
+  module RandomTools
+    # rubocop:disable MethodLength, Style/IfUnlessModifier
+
+    # Randomly choose an item among choices with integer weights. For example,
+    # if passed `{'a' => 3, 'b' => 1}`, the expected return value will approach
+    # 'a' 75% of the time and 'b' 25% of the time.
+    #
+    # @param [Hash{Object => Integer}] choices A mapping from each choice to an
+    #   integer weight. The weights will be relative to the sum of all weights.
+    #
+    # @return [Object] One of the keys chosen from the choices hash.
+    #
+    def self.random_weighted_sample(choices)
+      if choices.empty?
+        raise ArgumentError, 'Cannot choose among empty choices hash'
+      end
+
+      sum = 0
+      choices.each_pair do |item, weight|
+        unless weight.is_a?(Integer)
+          raise ArgumentError, "Choices must have integer weights, " \
+            "got #{item.inspect} => #{weight.inspect}"
+        end
+        sum += weight
+      end
+
+      target = rand(sum)
+
+      choices.each_pair do |item, weight|
+        return item if target < weight
+        target -= weight
+      end
+
+      raise NotImplementedError, 'This line should not be reached'
+    end
+
+    # rubocop:enable MethodLength, Style/IfUnlessModifier
+  end
+end
```

Comment:
Our code that chooses the AWS region for SES is based on the same principle, but we ended up using the simpler approach:
```ruby
region_pool = JSON.parse(Figaro.env.aws_ses_region_pool).flat_map do |region, weight|
  [region] * weight
end
region.sample

```
based on this article: http://www.rubyguides.com/2016/05/weighted-random-numbers

If you run this ruby file several times, you will see the results are quite similar, but the simpler approach has the advantage of being about 10x faster than the other one:
```ruby
require 'json'

SUBJECT = JSON.parse('{"a": 3, "b": 1}')

def sample_weighted
  adapters = SUBJECT.flat_map { |adapter, weight| [adapter] * weight }
  adapters.sample
end

def random_weighted
  max    = sum_of_weights
  target = rand(1..max)

  SUBJECT.each do |item, weight|
    return item if target <= weight
    target -= weight
  end
end

def sum_of_weights
  SUBJECT.inject(0) { |sum, (item, weight)| sum + weight }
end

counts = Hash.new(0)

1000.times { counts[random_weighted] += 1 }
puts "random_weighted: #{Float(counts["b"])/Float(counts["a"])}"

counts = Hash.new(0)

1000.times { counts[sample_weighted] += 1 }
puts "sample_weighted: #{Float(counts["b"])/Float(counts["a"])}"
```
See the discussion with benchmarks here: https://github.com/18F/identity-idp/pull/2038#pullrequestreview-102682277

Reply:
I don't think it's worth discussing performance differences that are < 1 ms.

```ruby
    def self.sample_weighted(choices)
      choices.each_pair do |item, weight|
        unless weight.is_a?(Integer)
          raise ArgumentError, "Choices must have integer weights, " \
            "got #{item.inspect} => #{weight.inspect}"
        end
      end

      weighted = choices.flat_map { |adapter, weight| [adapter] * weight }
      weighted.sample
    end
```

```ruby
#!/usr/bin/env ruby
require './lib/random_tools'
require 'benchmark'

choices = {a: 1, b: 1, c: 4, d: 3}
n = Integer(ARGV.fetch(0))

Benchmark.bm do |x|
   x.report('random_weighted_sample:') { n.times { Upaya::RandomTools.random_weighted_sample(choices) } }
   x.report('sample_weighted:') { n.times { Upaya::RandomTools.sample_weighted(choices) } }
end
```

```
ruby test.rb 1
                             user     system      total        real
random_weighted_sample:  0.000000   0.000000   0.000000 (  0.000029)
sample_weighted:         0.000000   0.000000   0.000000 (  0.000027)
```

```
ruby test.rb 1_000_000
                             user     system      total        real
random_weighted_sample:  1.120000   0.010000   1.130000 (  1.152417)
sample_weighted:         2.960000   0.040000   3.000000 (  3.058159)
```

Reply:
I stand corrected. My original benchmark from the other PR was flawed. The code that I proposed is actually about 2x slower than yours.